apple/internal/aspects/swift_static_framework_aspect.bzl

```diff
@@ -0,0 +1,152 @@
+    # Once the static framework rules are migrated to rules instead of macros, this aspect can be
+    # removed.
+    if ctx.rule.kind == "apple_static_library":
+        swiftdeps = [x for x in ctx.rule.attr.deps if SwiftInfo in x]
+
+        # If there are no Swift dependencies, return nothing.
+        if not swiftdeps:
+            return []
+
+        # If there's a different number of swift_library dependencies than all the declared
+        # dependencies, then there must be a mix of dependency types, which is not allowed for Swift
+        # based frameworks. We can't check that the count is exactly one since ctx.rule.attr.deps
+        # might return multiple configured targets for the same target based on the split.
+        # ctx.rule does not support split_attr, which seems to be an oversight. b/141931700 to track
+        # this.
+        if len(swiftdeps) != len(ctx.rule.attr.deps):
+            fail(
+                """\
+error: Found a mix of swift_library and other rule dependencies. Swift static frameworks expect a \
+single swift_library dependency with no transitive swift_library dependencies.\
+""",
+            )
+
+        # Collect all relevant artifacts for Swift static framework generation.
+        module_name = None
+        generated_header = None
+        swiftdocs = {}
+        swiftinterfaces = {}
+        for dep in swiftdeps:
+            swiftinfo = dep[SwiftInfo]
+
+            if len(swiftinfo.transitive_swiftinterfaces.to_list()) > 1:
+                fail(
+                    """\
+error: Found transitive swift_library dependencies. Swift static frameworks expect a single \
+swift_library dependency with no transitive swift_library dependencies.\
+""",
+                )
+
+            if not module_name:
+                module_name = swiftinfo.module_name
+            elif module_name != swiftinfo.module_name:
+                fail(
+                    """\
+error: Found multiple direct swift_library dependencies. Swift static frameworks expect a single \
+swift_library dependency with no transitive swift_library dependencies.\
+""",
+                )
+
+            arch = _swift_arch_for_dep(dep)
+
+            # Collect the interface artifacts. Only get the first element from each depset since
+            # they should only contain 1. If there are transitive swift_library dependencies, this
+            # aspect would have errored out before.
+            if not generated_header:
+                # Headers should be generated equally for all archs, so just take any of them.
+                generated_header = swiftinfo.transitive_generated_headers.to_list()[0]
+
+            swiftdocs[arch] = swiftinfo.transitive_swiftdocs.to_list()[0]
+            swiftinterfaces[arch] = swiftinfo.transitive_swiftinterfaces.to_list()[0]
+
+        # Make sure that all dictionaries contain at least one module before returning the provider.
+        if all([generated_header, swiftdocs, swiftinterfaces]):
```

Comment:
generated_header isn't required to exist if you set `swift.no_generated_header` because you never depend on the framework from objc, does that end up being required here?

Reply:
removed the headers being required